### PR TITLE
circulation: allow system librarian to update loan

### PR DESCRIPTION
**Pull request by zannkukai**
Logged as a system librarian, it's now possible to edit a loan, on a
patron circulation interface, to update the due date of this loan.

Co-Authored-by: Renaud Michotte <renaud.michotte@gmail.com>

## Dependencies

My PR depends on `rero-ils#<xx>`'s PR(s):

* https://github.com/rero/rero-ils/tree/US-cipo-reminders-fees

## How to test?

image

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: projects/admin/src/app/api/loan-api.service.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import { HttpClient, HttpHeaders } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+import { AbstractApiService } from './abstract-api.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class LoanApiService extends AbstractApiService{
```
Comment (Garfield-fr) [inline projects/admin/src/app/api/loan-api.service.ts:27]: Missing space after service and `{`

File: projects/admin/src/app/circulation/forms/update-loan-form/update-loan-form.component.html [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        <input type="text" class="form-control"
+               [ngClass]="{'is-invalid': formGroup.controls.endDate.touched && formGroup.controls.endDate.invalid}"
+               id="end-date" formControlName="endDate" aria-describedby="endDateHelp"
+               bsDatepicker required autocomplete="off"
+               [bsConfig]="bsConfig"
+        >
+      </div>
+      <ng-container *ngIf="formGroup.controls.endDate.touched && formGroup.controls.endDate.hasError('required')"
+                    [ngTemplateOutlet]="errorMessage"
+                    [ngTemplateOutletContext]="{message: 'The field is required' | translate}">
```
Comment (Garfield-fr) [inline projects/admin/src/app/circulation/forms/update-loan-form/update-loan-form.component.html:42]: Missing space after `{` and before `}`

File: projects/admin/src/app/circulation/forms/update-loan-form/update-loan-form.component.html [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  </div>
+  <div class="modal-body">
+    <div class="form-group">
+      <label for="end-date" class="label-title" translate>Due date</label>
+      <div class="input-group">
+        <div class="input-group-prepend">
+          <i class="fa fa-calendar input-group-text"></i>
+        </div>
+        <input type="text" class="form-control"
+               [ngClass]="{'is-invalid': formGroup.controls.endDate.touched && formGroup.controls.endDate.invalid}"
```
Comment (Garfield-fr) [inline projects/admin/src/app/circulation/forms/update-loan-form/update-loan-form.component.html:34]: Missing space after `{` and before `}`

File: projects/admin/src/app/api/loan-api.service.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  // CONSTRUCTOR ====================================================
+  /**
+   * Constructor
+   * @param _http - HttpClient
+   */
+  constructor(
+    private _http: HttpClient
+  ) {
+    super();
+  }
```
Comment (Garfield-fr) [inline projects/admin/src/app/api/loan-api.service.ts:40]: Remove this

File: projects/admin/src/app/circulation/forms/update-loan-form/update-loan-form.component.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+@Component({
+  selector: 'admin-update-loan-form',
+  templateUrl: './update-loan-form.component.html'
+})
+export class UpdateLoanFormComponent implements OnInit {
+
+  /** the date format to used */
+  static DATE_FORMAT = 'YYYY-MM-DD';
+  /** loan record metadata */
```
Comment (Garfield-fr) [inline projects/admin/src/app/circulation/forms/update-loan-form/update-loan-form.component.ts:36]: line before

File: projects/admin/src/app/circulation/item/item.component.html [hunk 6]
```diff
@@ -25,6 +25,12 @@
   <!-- ACTION BUTTONS -->
   <ng-container *ngIf="patron">
     <div class="actions">
+      <button class="btn btn-sm btn-outline-primary mr-1"
+              [attr.id]="item.barcode | idAttribute:{prefix: 'item', suffix: 'edit-loan-button'}"
```
Comment (Garfield-fr) [inline projects/admin/src/app/circulation/item/item.component.html:29]: Missing space after `{` and before `}`

File: projects/admin/src/app/circulation/item/item.component.html [hunk 7]
```diff
@@ -25,6 +25,12 @@
...
+              *ngIf="this.canEditLoan()"
+              (click)="openFixedEndDateDialog()">
+        <i class="fa fa-pencil"></i>
+      </button>
       <button class="btn btn-sm btn-outline-secondary"
               [disabled]="!item.actions.includes(itemAction.extend_loan)"
               [attr.id]="item.barcode | idAttribute:{prefix: 'item', suffix: 'renew-button'}"
```
Comment (Garfield-fr) [inline projects/admin/src/app/circulation/item/item.component.html:36]: Missing space after `{` and before `}`

File: projects/admin/src/app/circulation/item/item.component.ts [hunk 8]
```diff
@@ -66,27 +68,40 @@ export class ItemComponent implements OnInit {
...
+  }
+
+  // CONSTRUCTOR & HOOKS ============================================
   /**
    * Constructor
    * @param _recordService: Record Service
    * @param _organisationService: Organisation Service
    * @param _patronTransactionService: Patron transaction Service
    * @param _itemService: Item Service
+   * @param _userService: UserService
```
Comment (Garfield-fr) [inline projects/admin/src/app/circulation/item/item.component.ts:88]: add `@param _modalService: BsModalService`

**Review by Garfield-fr — APPROVED** [review]

File: projects/admin/src/app/api/loan-api.service.ts (hunk 4 above)
Comment (zannkukai) [reply projects/admin/src/app/api/loan-api.service.ts:40]: I can't because the parent doesn't (yet) declare the HttpClient

**Review by benerken — APPROVED** [review]

File: projects/admin/src/app/api/abstract-api.service.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
+/*
+ * RERO ILS UI
```
Comment (iGormilhit) [inline projects/admin/src/app/api/abstract-api.service.ts:2]: Why not *RERO ILS*?

**Review by iGormilhit — CHANGES_REQUESTED** [review]
Commit message title proposition: `circulation: allow system librarian to update loan` (without the capital initial, and a chararcter shorter :slightly_smiling_face: )

**Comment by zannkukai** [comment]
According with other developers, the direct Loan update isn't a good idea.

**Pull request closed without merging**